File: node/6_check_report.js
/*
 * 6_check_report.js  —  STEP 6: Get the score report
 * ===================================================
 * Checks ALL reports saved by Step 5 at the same time (concurrently).
 *   - Reports already "Completed" are skipped.
 *   - For the rest, the status is updated.
 *   - When a report is Completed, the full "details" block (download_url +
 *     expires_in_seconds) is saved.
 *   - Failed reports (or unreadable responses) are logged to errors.json,
 *     not data.json.
 *
 * EDIT NOTHING HERE. Your api_key lives in  ../config.json
 *
 * How to run:  node 6_check_report.js
 *
 * Note: the download_url expires after a short time (expires_in_seconds).
 * Download the score report PDF soon, or re-run this file.
 */

const { BASE_URL, apiKey, buildHeaders, getValue, saveValue, logJobError } = require("./helper");

// Statuses that mean "done, no need to check again".
const FINISHED = new Set(["completed"]);

// Reads the GET /report/{job_id} response.
// Returns [status, details, error].
function readReport(body) {
  if (body.success === false) {
    return ["Failed", null, body.error || {}];
  }
  const data = body.data || {};
  return [data.status, data.details, null];
}

async function checkOne(entry, headers) {
  const jobId = entry.job_id;
  let resp;
  try {
    resp = await fetch(`${BASE_URL}/report/${jobId}`, { headers });
  } catch (e) {
    console.log(`   - ${jobId}: request error (${e.message})`);
    logJobError(jobId, 0, "Request error: " + e.message, null);
    return false;
  }

  let body;
  try {
    body = await resp.json();
  } catch (e) {
    console.log(`   - ${jobId}: could not check (status code ${resp.status})`);
    logJobError(jobId, resp.status, "Could not read/parse report response", null);
    return false;
  }

  let [status, details, error] = readReport(body);
  status = status || "unknown";
  console.log(`   - ${jobId}: ${status}`);

  let changed = false;

  if (status !== entry.status) {
    entry.status = status;
    changed = true;
  }

  if (String(status).toLowerCase() === "completed" && details) {
    entry.details = details;
    changed = true;
    console.log(`     download_url: ${details.download_url}`);
    console.log(`     expires_in_seconds: ${details.expires_in_seconds}`);
  }

  if (error) {
    // Failed reports are not kept in data.json — they go to errors.json (job_errors).
    console.log(`     error: ${error.code} - ${error.detail}`);
    logJobError(jobId, resp.status, `${error.code || ""} ${error.detail || ""}`.trim(), error);
  }

  return changed;
}

async function main() {
  const key = apiKey();
  const reportProcess = getValue("report_process", []);

  if (reportProcess.length === 0) {
    console.log("[X] No reports found. Run 5_create_report.js first.");
    return;
  }

  const headers = buildHeaders(key);

  const pending = [];
  for (const entry of reportProcess) {
    if (FINISHED.has(String(entry.status || "").toLowerCase())) {
      console.log(`   - ${entry.job_id}: already ${entry.status} (skipped)`);
    } else {
      pending.push(entry);
    }
  }

  console.log(`\nChecking ${pending.length} report(s) concurrently...\n`);

  let changed = false;
  if (pending.length > 0) {
    const results = await Promise.all(pending.map((e) => checkOne(e, headers)));
    changed = results.some(Boolean);
  }

  if (changed) saveValue("report_process", reportProcess);

  const done = reportProcess.filter((r) =>
    FINISHED.has(String(r.status || "").toLowerCase())
  );
  const stillPending = reportProcess.filter(
    (r) => !FINISHED.has(String(r.status || "").toLowerCase())
  );

  console.log("\nSummary:");
  console.log(`   finished: ${done.length}  |  still generating: ${stillPending.length}`);

  if (stillPending.length > 0) {
    console.log("Some reports are still generating. Wait a moment and run this file again.");
  } else {
    console.log(
      "[OK] All reports finished. Download your score report PDF(s) using the URL(s) above."
    );
  }
}

main();

File: python-async/6_check_report.py
"""
6_check_report.py (async)  —  STEP 6: Get the score report
==========================================================
Checks ALL reports saved by Step 5 at the same time (concurrently).
  - Reports already "Completed" are skipped.
  - When a report is Completed, the full "details" block is saved.
  - Failed reports (or unreadable responses) are logged to errors.json, not data.json.

EDIT NOTHING HERE. Your api_key lives in  ../config.json

How to run:  python 6_check_report.py

Note: the download_url expires after a short time (expires_in_seconds).
Download the score report PDF soon, or re-run this file.
"""

import asyncio
import httpx
from helper import BASE_URL, api_key, build_headers, get_value, save_value, log_job_error

FINISHED = {"completed"}


def read_report(body):
    if body.get("success") is False:
        return "Failed", None, body.get("error") or {}
    data = body.get("data") or {}
    return data.get("status"), data.get("details"), None


async def check_one(client, entry, headers):
    job_id = entry.get("job_id")
    try:
        resp = await client.get(f"{BASE_URL}/report/{job_id}", headers=headers)
    except httpx.HTTPError as e:
        print(f"   - {job_id}: request error ({e})")
        log_job_error(job_id, 0, f"Request error: {e}", None)
        return False

    try:
        status, details, error = read_report(resp.json())
    except ValueError:
        print(f"   - {job_id}: could not check (status code {resp.status_code})")
        log_job_error(job_id, resp.status_code, "Could not read/parse report response", None)
        return False

    status = status or "unknown"
    print(f"   - {job_id}: {status}")

    changed = False

    if status != entry.get("status"):
        entry["status"] = status
        changed = True

    if str(status).lower() == "completed" and details:
        entry["details"] = details
        changed = True
        print(f"     download_url: {details.get('download_url')}")
        print(f"     expires_in_seconds: {details.get('expires_in_seconds')}")

    if error:
        # Failed reports are not kept in data.json — they go to errors.json (job_errors).
        code = error.get("code", "")
        detail = error.get("detail", "")
        print(f"     error: {code} - {detail}")
        log_job_error(job_id, resp.status_code, f"{code} {detail}".strip(), error)

    return changed


async def main():
    key = api_key()
    report_process = get_value("report_process", [])

    if not report_process:
        print("[X] No reports found. Run 5_create_report.py first.")
        return

    headers = build_headers(key)

    pending_entries = []
    for entry in report_process:
        if str(entry.get("status", "")).lower() in FINISHED:
            print(f"   - {entry.get('job_id')}: already {entry.get('status')} (skipped)")
        else:
            pending_entries.append(entry)

    print(f"\nChecking {len(pending_entries)} report(s) concurrently...\n")

    changed = False
    if pending_entries:
        async with httpx.AsyncClient() as client:
            results = await asyncio.gather(
                *[check_one(client, e, headers) for e in pending_entries]
            )
        changed = any(results)

    if changed:
        save_value("report_process", report_process)

    done = [r for r in report_process if str(r.get("status", "")).lower() in FINISHED]
    pending = [r for r in report_process if str(r.get("status", "")).lower() not in FINISHED]

    print("\nSummary:")
    print(f"   finished: {len(done)}  |  still generating: {len(pending)}")

    if pending:
        print("Some reports are still generating. Wait a moment and run this file again.")
    else:
        print("[OK] All reports finished. Download your score report PDF(s) using the URL(s) above.")


if __name__ == "__main__":
    asyncio.run(main())
